src/module.c

```diff
@@ -10846,10 +10998,407 @@ sds genModulesInfoString(sds info) {
+            return REDISMODULE_ERR;
+        }
+        sdsfreesplitres(config_pair, count);
+        listDelNode(server.module_configs_queue, ln);
+    }
+    return REDISMODULE_OK;
+}
+
+/* Add a moduleConfigTuple to the list if the apply and privdata do not match one already in it. */
+void addModuleApply(list *module_configs_tuples, void *module_config_val) {
+    ModuleConfig *module_config = module_config_val;
+    if (!module_config->apply_fn) return;
+    listIter li;
+    listNode *ln;
+    moduleConfigTuple *module_config_tuple;
+    ModuleConfig *pending_apply;
+    listRewind(module_configs_tuples, &li);
+    while ((ln = listNext(&li))) {
+        module_config_tuple = ln->value;
+        pending_apply = module_config_tuple->module_config;
+        if (pending_apply->apply_fn == module_config->apply_fn && pending_apply->privdata == module_config->privdata) {
+            return;
+        }
+    }
+    moduleConfigTuple *tuple = zmalloc(sizeof(moduleConfigTuple));
+    tuple->module_config = module_config;
+    tuple->module = module_config->module;
+    listAddNodeTail(module_configs_tuples, tuple);
+}
+
+/* Call apply on all module configs specified in set, if an apply function was specified at registration time. */
+int moduleConfigApplyConfig(list *module_configs_tuples, const char **err, const char **err_arg_name) {
+    if (!listLength(module_configs_tuples)) return 1;
+    listIter li;
+    listNode *ln;
+    moduleConfigTuple *module_config_tuple;
+    ModuleConfig *module_config;
+    RedisModuleCtx ctx;
+
+    listRewind(module_configs_tuples, &li);
+    while ((ln = listNext(&li))) {
+        module_config_tuple = ln->value;
+        module_config = module_config_tuple->module_config;
+        moduleCreateContext(&ctx, module_config_tuple->module, REDISMODULE_CTX_NONE);
+        if (module_config->apply_fn(&ctx, module_config->privdata, err)) {
+            if (err_arg_name) *err_arg_name = module_config->name;
+            return 0;
+        }
+        moduleFreeContext(&ctx);
+    }
+    return 1;
+}
+
+/* --------------------------------------------------------------------------
+ * ## Module Configurations API
+ * -------------------------------------------------------------------------- */
+
+#define createModuleConfig(config_name, config_type, apply, config_privdata, module) { \
+    .name = (config_name), \
+    .type = (config_type), \
+    .privdata = (config_privdata), \
+    .apply_fn = (apply), \
+    .module = (module) \
+}
+
+#define addDefaultAndCallbacksToConfig(config, getfn, setfn, type) \
+    config.get_fn.get_##type = getfn; \
+    config.set_fn.set_##type = setfn;
+
+void addModuleConfigtoList(RedisModule *module, ModuleConfig *new_module_config) {
+    ModuleConfig *new_config = zmalloc(sizeof(ModuleConfig));
+    memcpy(new_config, new_module_config, sizeof(ModuleConfig));
+    module->module_configs = listAddNodeTail(module->module_configs, new_config);
+}
```

Comment:
This being a macro is a bit weird and unexpected, I would also expect this to actually return a type.
```suggestion
// The types on this aren't right, but I think you get the idea
ModuleConfig *createModuleConfig(config_name, config_type, apply, config_privdata, module) {
    ModuleConfig *new_config = zmalloc(sizeof(ModuleConfig);
    new_config->name = config_name;
    new_config->type = config_type;
    new_config->privdata = privdata;
    new_config->apply_fn = apply;
    new_config->module = module;
    return new_config;
}

#define addDefaultAndCallbacksToConfig(config, getfn, setfn, type) \
    config->get_fn.get_##type = getfn; \
    config->set_fn.set_##type = setfn;

void addModuleConfigtoList(RedisModule *module, ModuleConfig *new_config) {
    module->module_configs = listAddNodeTail(module->module_configs, new_config);
}
```